src/xtopcom/xstatistic/src/xstatistic.cpp

```diff
@@ -0,0 +1,167 @@
+        }
+    }
+    return m_size;
+}
+
+void xobject_statistic_base_t::add_object(xstatistic_obj_face_t * object) {
+    std::lock_guard<std::mutex> lck(m_mutex);
+    refresh_inner(object->create_time());
+#ifdef USE_MULTISET_ONLY
+    m_not_calc_object_set.insert(object);
+#else
+    auto iter = m_not_calc_object_set.insert(object);
+    m_not_calc_object_map[object] = iter;
+#endif
+    xdbg("xobject_statistic_base_t::add_object object:%p, type:%d", object, object->get_class_type());
+}
+
+void xobject_statistic_base_t::del_object(xstatistic_obj_face_t * object) {
+    std::lock_guard<std::mutex> lck(m_mutex);
+#ifdef USE_MULTISET_ONLY
+    auto ret = m_not_calc_object_set.equal_range(object);
+    for (auto it = ret.first; it != ret.second; it++) {
+        xdbg("xobject_statistic_base_t::del_object it:%p, object:%p, type:%d", (*it), object, object->get_class_type());
+        if ((*it) == object) {
+            xdbg("xobject_statistic_base_t::del_object erase object:%p, type:%d", object, object->get_class_type());
+            m_not_calc_object_set.erase(it);
+            refresh_inner(base::xtime_utl::gmttime_ms());
+            return;
+        }
+    }
+
+    update_metrics(object->get_class_type(), -1, -object->get_object_size());
+    refresh_inner(base::xtime_utl::gmttime_ms());
+#else
+    auto it_map = m_not_calc_object_map.find(object);
+    if (it_map != m_not_calc_object_map.end()) {
+        xdbg("xobject_statistic_base_t::del_object from map object:%p, type:%d", object, object->get_class_type());
+        m_not_calc_object_set.erase(it_map->second);
+        m_not_calc_object_map.erase(it_map);
+    } else {
+        xdbg("xobject_statistic_base_t::del_object del size object:%p, type:%d", object, object->get_class_type());
+        update_metrics(object->get_class_type(), -1, -object->get_object_size());
+    }
+    
+    refresh_inner(base::xtime_utl::gmttime_ms());
+#endif
+}
+
+void xobject_statistic_base_t::refresh() {
+    auto now = base::xtime_utl::gmttime_ms();
+    std::lock_guard<std::mutex> lck(m_mutex);
+    refresh_inner(now);
+}
+
+void xobject_statistic_base_t::refresh_inner(int64_t now) {
+    for (auto iter = m_not_calc_object_set.begin(); iter != m_not_calc_object_set.end();) {
+        xdbg("xobject_statistic_base_t::refresh_inner object:%p, type:%d, create time:%llu, delay_time:%d, now:%llu",
+             (*iter),
+             (*iter)->get_class_type(),
+             (*iter)->create_time(),
+             STATISTIC_DELAY_TIME,
+             now);
+        if ((*iter)->create_time() + STATISTIC_DELAY_TIME <= now) {
+            auto size = (*iter)->get_object_size();
+            update_metrics((*iter)->get_class_type(), 1, size);
+            xdbg("xobject_statistic_base_t::refresh_inner object:%p, size:%d, type:%d", (*iter), size, (*iter)->get_class_type());
+#ifndef USE_MULTISET_ONLY
+            m_not_calc_object_map.erase(*iter);

```

Comment:
iter doesn't need update?

Reply:
yes, iter belongs to m_not_calc_object_set, not m_not_calc_object_map.